### PR TITLE
feat: web sdk topics

### PR DESCRIPTION
**Pull request by pgautier404**
This commit adds support for publish and subscribe to the Web SDK and refactors the topic client integration tests so they are shared between the nodejs and web SDKs.

### CODE REVIEW

File: packages/client-sdk-web/src/internal/pubsub-client.ts [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+{
+  private readonly client: pubsub.PubsubClient;
+  private readonly configuration: Configuration;
+  private readonly credentialProvider: CredentialProvider;
+  // private readonly unaryRequestTimeoutMs: number;
+  // private static readonly DEFAULT_REQUEST_TIMEOUT_MS: number = 5 * 1000;
+  private readonly logger: MomentoLogger;
+  private readonly authHeaders: {authorization: string};
+  private readonly unaryInterceptors: UnaryInterceptor<REQ, RESP>[];
+  private readonly streamingInterceptors: StreamInterceptor<REQ, RESP>[];
```
Comment (malandis) [inline packages/client-sdk-web/src/internal/pubsub-client.ts:83]: Are these a strictly gRPC-web feature? A zoom call could help me understand this part.

**Review by cprice404 — APPROVED** [review]
awesome!  lgtm but let's have @malandis review this since he did the original work.

happening

File: packages/client-sdk-web/src/internal/pubsub-client.ts [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+export class PubsubClient<
+  REQ extends Request<REQ, RESP>,
+  RESP extends UnaryResponse<REQ, RESP>
+> implements IPubsubClient
+{
+  private readonly client: pubsub.PubsubClient;
+  private readonly configuration: Configuration;
+  private readonly credentialProvider: CredentialProvider;
+  // private readonly unaryRequestTimeoutMs: number;
+  // private static readonly DEFAULT_REQUEST_TIMEOUT_MS: number = 5 * 1000;
```
[the remark this answers is not in the document]
Comment (pgautier404) [reply packages/client-sdk-web/src/internal/pubsub-client.ts:79]: Yes, all remaining comments are waiting on config plumbing.

**Review by malandis — COMMENTED** [review]
Looks great. I have a couple questions below about how much code reuse is possible. I see the interfaces were introduced for the benefit of shared tests, which is really nice!

File: packages/client-sdk-web/src/internal/grpc/headers-interceptor.ts [hunk 3]
```diff
@@ -45,12 +51,34 @@ export class HeaderInterceptorProvider<
       ): Promise<UnaryResponse<REQ, RESP>> => {
         const md = request.getMetadata();
         this.headersToAddEveryTime.forEach(h => (md[h.name] = h.value));
-        if (HeaderInterceptorProvider.areOnlyOnceHeadersSent) {
+        if (!HeaderInterceptorProvider.areOnlyOnceHeadersSent) {
```
[the remark this answers is not in the document]
Comment (pgautier404) [reply packages/client-sdk-web/src/internal/grpc/headers-interceptor.ts:54]: I'm not sure how it slipped past previously. I just noticed it didn't smell right and fixed it.

File: packages/client-sdk-web/src/internal/grpc/headers-interceptor.ts [hunk 4]
```diff
@@ -45,12 +51,34 @@ export class HeaderInterceptorProvider<
...
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const self = this;
+    return {
+      intercept(
+        request: Request<REQ, RESP>,
+        invoker: (request: Request<REQ, RESP>) => ClientReadableStream<RESP>
+      ): ClientReadableStream<RESP> {
+        const md = request.getMetadata();
+        self.headersToAddEveryTime.forEach(h => (md[h.name] = h.value));
+        if (!HeaderInterceptorProvider.areOnlyOnceHeadersSent) {
```
[the remark this answers is not in the document]
Comment (pgautier404) [reply packages/client-sdk-web/src/internal/grpc/headers-interceptor.ts:76]: Great point-- I'll take a closer look.

File: packages/client-sdk-web/src/topic-client-props.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import {CredentialProvider} from '.';
+import {Configuration} from './config/configuration';
+
+export interface TopicClientProps {
```
[the remark this answers is not in the document]
Comment (pgautier404) [reply packages/client-sdk-web/src/topic-client-props.ts:4]: Not yet. We need to wait until Configuration is plumbed through on the Web SDK side before it (and this) can be moved to core.

File: packages/client-sdk-web/src/internal/grpc/headers-interceptor.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (pgautier404) [reply packages/client-sdk-web/src/internal/grpc/headers-interceptor.ts]: Fixed in 7dbf591.

**Review by malandis — APPROVED** [review]

**Pull request merged**